src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statehistory = [];
+  const stateForEachAction = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateForEachAction, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete stateForEachAction[key];
+        }
+        break;
+
+      case 'clear':
+        for (const key in stateForEachAction) {
+          delete stateForEachAction[key];
+        }
+        break;
+    }
```

Comment:
There should be a `default` case in your `switch` statement to handle any unexpected `action.type` values.